File: integrations/pydantic-ai/python/examples/basic_agent.py
"""
Basic Pydantic AI agent with Zep long-term memory.

This example wires Zep into a Pydantic AI agent using three pieces:

  - ``create_user`` /
    ``create_thread``         -- create the Zep resources one time and return
                                 the server-generated UUIDs.
  - ``ZepDeps``               -- carries the Zep client + the user, thread, and
                                 graph UUIDs.
  - ``zep_capabilities``      -- bundles the history processor (persists each
                                 user turn and injects Zep's context block)
                                 with automatic assistant-reply persistence.
  - ``create_zep_search_tool``-- a model-callable graph-search tool.

The flow each turn:

  1. The history processor persists the user's message to Zep and prepends the
     retrieved context block to the prompt.
  2. The model answers (optionally calling ``zep_search``).
  3. The ``after_run`` hook (bundled by ``zep_capabilities``) automatically
     persists the assistant's reply back to the Zep thread -- no explicit
     ``persist_run`` call needed.

Prerequisites:
    pip install zep-pydantic-ai

    export ZEP_API_KEY="your-zep-api-key"
    export OPENAI_API_KEY="your-openai-api-key"
"""

from __future__ import annotations

import asyncio
import os
import sys
import time

from pydantic_ai import Agent
from zep_cloud.client import AsyncZep

from zep_pydantic_ai import (
    ZepDeps,
    create_thread,
    create_user,
    create_zep_search_tool,
    zep_capabilities,
)

ZEP_API_KEY = os.environ.get("ZEP_API_KEY", "")
OPENAI_API_KEY = os.environ.get("OPENAI_API_KEY", "")

if not ZEP_API_KEY:
    raise OSError("ZEP_API_KEY is not set.")
if not OPENAI_API_KEY:
    raise OSError("OPENAI_API_KEY is not set.")


async def wait_for_ingestion(
    zep: AsyncZep,
    graph_uuid: str,
    *,
    minimum_episodes: int = 2,
    timeout_seconds: int = 180,
    poll_seconds: float = 5.0,
) -> None:
    """Poll the graph until its episodes are processed, or the timeout ends.

    Zep ingests asynchronously, so a fact is not retrievable immediately.
    """
    deadline = time.monotonic() + timeout_seconds
    while time.monotonic() < deadline:
        pager = await zep.graph.episode.list(graph_uuid, limit=20)
        episodes = pager.items or []
        processed = [episode for episode in episodes if episode.processed]
        if len(processed) >= minimum_episodes:
            print(f"  {len(processed)} episodes are processed.\n")
            return
        await asyncio.sleep(poll_seconds)
    print("  The timeout ended before all episodes were processed.\n")


async def chat(agent: Agent, deps: ZepDeps, message: str) -> str:
    """Send one message through the agent; the assistant reply is persisted
    automatically by the after_run hook bundled in zep_capabilities(deps)."""
    result = await agent.run(message, deps=deps)
    return result.output


async def main() -> None:
    zep = AsyncZep(api_key=ZEP_API_KEY)

    # Zep v4 addresses every resource by a server-generated UUID. A real
    # application creates the user and the thread one time, and stores the
    # UUIDs in its own database.
    #
    # Note: ZEPAI-3605 -- a user that is created without a ``user_id`` cannot
    # receive a thread message until the fix is deployed.
    user = await create_user(
        zep,
        first_name="Alice",
        last_name="Smith",
        email="alice@example.com",
    )
    thread = await create_thread(zep, user_uuid=user.uuid_)

    # ZepDeps must exist before the agent: zep_capabilities(deps) closes over
    # it to wire up automatic assistant persistence via Hooks(after_run=...).
    deps = ZepDeps(
        client=zep,
        user_uuid=user.uuid_,
        thread_uuid=thread.uuid_,
        graph_uuid=user.graph_uuid,
        first_name="Alice",
        last_name="Smith",
    )

    agent = Agent(
        "openai:gpt-5-mini",
        deps_type=ZepDeps,
        capabilities=zep_capabilities(deps),
        tools=[create_zep_search_tool()],
        instructions=(
            "You are a helpful assistant with access to long-term memory. "
            "When context from Zep is injected into the prompt, use it to provide "
            "personalised, memory-aware responses. If you know something about the "
            "user from memory, reference it naturally. Use the zep_search tool when "
            "you need to look up specific details the user shared earlier."
        ),
    )

    print(f"\n{'=' * 60}")
    print("Pydantic AI + Zep Memory Example")
    print(f"{'=' * 60}")
    print(f"  User UUID:   {user.uuid_}")
    print(f"  Thread UUID: {thread.uuid_}")
    print(f"  Graph UUID:  {user.graph_uuid}")
    print(f"{'=' * 60}\n")

    # Phase 1: seed some facts.
    print("--- Phase 1: Seeding facts ---\n")
    for msg in (
        "My name is Alice and I'm a software engineer.",
        "I live in Portland and love hiking on weekends.",
    ):
        print(f"User:  {msg}")
        print(f"Agent: {await chat(agent, deps, msg)}\n")

    # Phase 2: let Zep's asynchronous ingestion build the graph.
    print("--- Phase 2: Waiting for Zep graph processing ---\n")
    if user.graph_uuid:
        await wait_for_ingestion(zep, user.graph_uuid)

    # Phase 3: test memory recall in the same thread.
    print("--- Phase 3: Testing memory recall ---\n")
    checks = (
        ("What do I do for work?", ("engineer", "software")),
        ("Where do I live?", ("portland",)),
    )
    failures = []
    for msg, expected in checks:
        answer = await chat(agent, deps, msg)
        print(f"User:  {msg}")
        print(f"Agent: {answer}\n")
        if not any(word in answer.lower() for word in expected):
            failures.append(f"{msg!r} did not recall any of {expected}")

    # Phase 4: clean up the resources that this example created.
    print("--- Phase 4: Cleanup ---\n")
    await zep.user.delete(user.uuid_)

    if failures:
        for failure in failures:
            print(f"FAIL: {failure}")
        sys.exit(1)

    print("Done!")


if __name__ == "__main__":
    asyncio.run(main())

File: integrations/mastra/typescript/examples/basic-agent.ts
/**
 * Basic Mastra agent with automatic Zep long-term memory.
 *
 * Demonstrates the full loop:
 *   1. Create a Zep user + thread and read their server-generated UUIDs
 *      (`createZepUserAndThread`).
 *   2. Build the Zep input/output processor pair (`createZepProcessors`).
 *   3. Attach the processors to a Mastra `Agent` via `inputProcessors` /
 *      `outputProcessors` — no tool-calling round-trip needed.
 *   4. Seed facts across turns (persisted automatically by the output
 *      processor), wait for asynchronous ingestion, then ask a question that
 *      requires recalling them (answered using context the input processor
 *      injects automatically).
 *
 * Known defect (ZEPAI-3605): a user that is created without a `userId` cannot
 * receive a thread message until the fix is deployed. The output processor
 * reports the failure and the agent continues.
 *
 * Prerequisites:
 *   npm install
 *   export ZEP_API_KEY="your-zep-api-key"
 *   export OPENAI_API_KEY="your-openai-api-key"
 *
 * Run:
 *   npm run example
 */

import { ZepClient } from "@getzep/zep-cloud";
import { Agent } from "@mastra/core/agent";
import { createZepProcessors, createZepUserAndThread } from "../src/index.js";

const ZEP_API_KEY = process.env.ZEP_API_KEY;
const OPENAI_API_KEY = process.env.OPENAI_API_KEY;

if (!ZEP_API_KEY) throw new Error("ZEP_API_KEY is not set.");
if (!OPENAI_API_KEY) throw new Error("OPENAI_API_KEY is not set.");

async function ask(agent: Agent, prompt: string): Promise<string> {
  console.log(`\nUser:  ${prompt}`);
  const result = await agent.generate(prompt);
  console.log(`Agent: ${result.text}`);
  return result.text;
}

async function main(): Promise<void> {
  const client = new ZepClient({ apiKey: ZEP_API_KEY });

  console.log("=".repeat(60));
  console.log("Mastra + Zep Automatic Memory Example");
  console.log("=".repeat(60));

  // 1. Create the user and the thread before the first turn. Zep generates
  //    the UUIDs; a real application stores them in its own database.
  const identity = await createZepUserAndThread({
    client,
    firstName: "Alice",
    lastName: "Smith",
    email: "alice@example.com",
  });
  if (!identity) throw new Error("Could not create the Zep user and thread.");
  console.log(`  User UUID:   ${identity.userUuid}`);
  console.log(`  Graph UUID:  ${identity.graphUuid}`);
  console.log(`  Thread UUID: ${identity.threadUuid}`);

  // 2. Build the Zep input/output processor pair bound to these UUIDs.
  const { inputProcessor, outputProcessor } = createZepProcessors({
    client,
    graphUuid: identity.graphUuid,
    threadUuid: identity.threadUuid,
  });

  // 3. Attach to a Mastra agent (id AND name are both required). No tools
  //    needed — the processors persist and inject context automatically.
  const agent = new Agent({
    id: "memory-agent",
    name: "Memory Agent",
    instructions:
      "You are a helpful assistant with long-term memory about the user, " +
      "injected automatically into your context. Personalize your replies " +
      "using what you know about the user.",
    model: "openai/gpt-5-mini",
    inputProcessors: [inputProcessor],
    outputProcessors: [outputProcessor],
  });

  // 4a. Seed facts — the output processor persists each turn automatically.
  console.log("\n--- Phase 1: Seeding facts ---");
  await ask(agent, "Hi! I live in Portland and love hiking.");
  await ask(agent, "I also work as a software engineer.");

  // 4b. Wait for asynchronous graph ingestion.
  const waitSeconds = 15;
  console.log(`\n--- Waiting ${waitSeconds}s for Zep graph processing ---`);
  await new Promise((resolve) => setTimeout(resolve, waitSeconds * 1000));

  // 4c. Recall — the input processor injects the Context Block automatically.
  console.log("\n--- Phase 2: Recall ---");
  await ask(agent, "What do you remember about where I live and what I do?");

  console.log("\nDone.");
}

main().catch((error) => {
  console.error("Example failed:", error);
  process.exitCode = 1;
});
